apps/aechannel/src/aesc_fsm.erl

```diff
@@ -625,82 +642,112 @@ awaiting_signature(cast, {?SIGNED, slash_tx, SignedTx} = Msg,
 awaiting_signature(cast, {?SIGNED, ?UPDATE, SignedTx} = Msg, D) ->
-    D1 = send_update_msg(
-           SignedTx,
-           D#data{state = aesc_offchain_state:set_half_signed_tx(
-                            SignedTx, D#data.state)}),
-    next_state(awaiting_update_ack,
-               log(rcv, ?SIGNED, Msg, D1#data{latest = undefined}));
+    maybe_check_sigs(SignedTx, channel_offchain_tx, not_offchain_tx, me,
+        fun() ->
+            D1 = send_update_msg(
+                  SignedTx,
+                  D#data{state = aesc_offchain_state:set_half_signed_tx(
+                                    SignedTx, D#data.state)}),
+            next_state(awaiting_update_ack,
+                      log(rcv, ?SIGNED, Msg, D1#data{latest = undefined}))
+        end, D);
 awaiting_signature(cast, {?SIGNED, ?UPDATE_ACK, SignedTx} = Msg,
                    #data{latest = {sign, ?UPDATE_ACK, OCTx}, opts = Opts} = D) ->
     NewSignedTx = aetx_sign:add_signatures(
                     OCTx, aetx_sign:signatures(SignedTx)),
-    D1 = send_update_ack_msg(NewSignedTx, D),
-    {OnChainEnv, OnChainTrees} =
-        aetx_env:tx_env_and_trees_from_top(aetx_contract),
-    State = aesc_offchain_state:set_signed_tx(NewSignedTx, D1#data.state,
-                                              OnChainTrees, OnChainEnv, Opts),
-    D2 = D1#data{log   = log_msg(rcv, ?SIGNED, Msg, D1#data.log),
-                 state = State,
-                 latest = undefined},
-    next_state(open, D2);
+    maybe_check_sigs(NewSignedTx, channel_offchain_tx, not_offchain_tx, both,
+        fun() ->
+            D1 = send_update_ack_msg(NewSignedTx, D),
+            {OnChainEnv, OnChainTrees} =
+                aetx_env:tx_env_and_trees_from_top(aetx_contract),
+            State = aesc_offchain_state:set_signed_tx(NewSignedTx, D1#data.state,
+                                                      OnChainTrees, OnChainEnv, Opts),
+            D2 = D1#data{log   = log_msg(rcv, ?SIGNED, Msg, D1#data.log),
+                        state = State,
+                        latest = undefined},
+            next_state(open, D2)
+        end, D);
 awaiting_signature(cast, {?SIGNED, ?SHUTDOWN, SignedTx} = Msg, D) ->
-    D1 = send_shutdown_msg(SignedTx, D),
-    D2 = D1#data{latest = {shutdown, SignedTx},
-                 log    = log_msg(rcv, ?SIGNED, Msg, D1#data.log)},
-    next_state(mutual_closing, D2);
+    maybe_check_sigs(SignedTx, channel_close_mutual_tx, not_close_mutual_tx, me,
+        fun() ->
+            D1 = send_shutdown_msg(SignedTx, D),
+            D2 = D1#data{latest = {shutdown, SignedTx},
+                        log    = log_msg(rcv, ?SIGNED, Msg, D1#data.log)},
+            next_state(mutual_closing, D2)
+        end, D);
 awaiting_signature(cast, {?SIGNED, ?SHUTDOWN_ACK, SignedTx} = Msg,
                    #data{latest = {sign, ?SHUTDOWN_ACK, CMTx}} = D) ->
     NewSignedTx = aetx_sign:add_signatures(
                     CMTx, aetx_sign:signatures(SignedTx)),
```

Comment:
My eyes hurt from all the missing abstraction in this code, but that has nothing to do with the merge...